pubsub/psclient/client.go

```diff
@@ -0,0 +1,217 @@
+package client
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	exptypes "github.com/decred/dcrdata/v4/explorer/types"
+	pstypes "github.com/decred/dcrdata/v4/pubsub/types"
+	"golang.org/x/net/websocket"
+)
+
+// newSubscribeMsg creates a new subscribe message equivalent to marshaling a
+// pstypes.WebSocketMessage with EventId set to "subscribe" and Message set to
+// the input string, event. json.Marshal is not used, but tests ensure the
+// correct result.
+func newSubscribeMsg(event string) string {
+	return fmt.Sprintf(`{"event":"subscribe","message":"%s"}`, event)
+}
+
+// newUnsubscribeMsg creates a new unsubscribe message equivalent to marshaling
+// a pstypes.WebSocketMessage with EventId set to "unsubscribe" and Message set
+// to the input string, event. json.Marshal is not used, but tests ensure the
+// correct result.
+func newUnsubscribeMsg(event string) string {
+	return fmt.Sprintf(`{"event":"unsubscribe","message":"%s"}`, event)
+}
+
+var defaultTimeout = 10 * time.Second
+
+// Client wraps a *websocket.Conn.
+type Client struct {
+	*websocket.Conn
+	ReadTimeout  time.Duration
+	WriteTimeout time.Duration
+}
+
+// New creates a new Client from a *websocket.Conn.
+func New(ws *websocket.Conn) *Client {
+	if ws == nil {
+		return nil
+	}
+	return &Client{
+		Conn:         ws,
+		ReadTimeout:  defaultTimeout,
+		WriteTimeout: defaultTimeout,
+	}
+}
+
+// Subscribe sends a subscribe type WebSocketMessage for the given event name
+// after validating it. The response is returned.
+func (c *Client) Subscribe(event string) (*pstypes.WebSocketMessage, error) {
+	// Validate the event type.
+	_, ok := pstypes.Subscriptions[event]
+	if !ok {
+		return nil, fmt.Errorf("invalid subscription %s", event)
+	}
+
+	// Send the subscribe message.
+	msg := newSubscribeMsg(event)
+	_ = c.SetWriteDeadline(time.Now().Add(c.WriteTimeout))
+	_, err := c.Write([]byte(msg))
+	if err != nil {
+		return nil, fmt.Errorf("failed to send subscribe message: %v", err)
+	}
+
+	// Read the response.
+	return c.ReceiveMsg()
```

Comment:
Running Subscribe also reads from the websocket.Conn, so there should not be readers in concurrent goroutines.  Perhaps the ReceiveMsg here could be omitted to allow another ready handle it.